torch/onnx/_internal/exporter.py

```diff
@@ -0,0 +1,248 @@
+import logging
+from typing import (
+    Any,
+    Callable,
+    Final,
+    Mapping,
+    Optional,
+    Protocol,
+    runtime_checkable,
+    Sequence,
+    TYPE_CHECKING,
+    TypeVar,
+    Union,
+)
+
+import torch
+
+from torch.onnx._internal import _beartype
+
+# We can only import onnx from this module in a type-checking context to ensure that
+# 'import torch.onnx' continues to work without having 'onnx' installed. We fully
+# 'import onnx' inside of dynamo_export (by way of _assert_dependencies).
+if TYPE_CHECKING:
+    import onnx
+
+
+_DEFAULT_OPSET_VERSION: Final[int] = 18
+"""The default ONNX opset version the exporter will use if one is not specified explicitly
+through ``ExportOptions``. This should NEVER be accessed outside of this module! Users
+should reference ``ExportOptions.opset_version``."""
+
+
+@_beartype.beartype
+class ExportOptions:
+    """Options to influence the TorchDynamo ONNX exporter."""
+
+    opset_version: Optional[int] = None
+    """The ONNX opset version the exporter should target. Defaults to the latest
+    supported ONNX opset version. The default version will increment over time as
+    ONNX continues to evolve."""
+
+    dynamic_shapes: Optional[bool] = None
+    """Shape information hint for input/output tensors.
+
+    - ``None``: the exporter determines the most compatible setting.
+    - ``True``: all input shapes are considered dynamic.
+    - ``False``: all input shapes are considered static."""
+
+    logger: Optional[logging.Logger] = None
+    """The logger for the ONNX exporter to use. Defaults to creating a child
+    logger named "torch.onnx" under the current logger (as returned by
+    :py:meth:`logging.getLogger`)."""
+
+    def __init__(
+        self,
+        *,
+        opset_version: Optional[int] = None,
+        dynamic_shapes: Optional[bool] = None,
+        logger: Optional[logging.Logger] = None,
+    ):
+        self.opset_version = opset_version
+        self.dynamic_shapes = dynamic_shapes
+        self.logger = logger
+
+
+@runtime_checkable
+class ExportOutputSerializer(Protocol):
+    """Protocol for serializing an ONNX graph into a specific format (e.g. Protobuf)."""
+
+    def serialize(
+        self, export_output: ExportOutput, destination: io.BufferedIOBase
+    ) -> None:
+        ...
+
+
+class ProtobufExportOutputSerializer:
+    """Serializes ONNX graph as Protobuf."""
+
+    @_beartype.beartype
+    def serialize(
+        self, export_output: ExportOutput, destination: io.BufferedIOBase
+    ) -> None:
+        import onnx
+
+        if not isinstance(export_output.model_proto, onnx.ModelProto):
+            raise ValueError("export_output.ModelProto is not an onnx.ModelProto")
+        destination.write(export_output.model_proto.SerializeToString())
+
+
+class ExportOutput:
+    """An in-memory representation of a PyTorch model that has been exported to ONNX."""
+
+    _export_options: Final[ExportOptions]
+    _model_proto: Final[onnx.ModelProto]
```

Comment:
```suggestion
    _model_proto: Final[onnx.ModelProto]
    """Model proto used by ONNX spec to represent the model.
    
    NOTE: In future releases this will be replaced by an agnostic serialization graph IR to allow multiple serialization formats.
    """
    Maybe a note on why `ExportOutput` has a model_proto instead of it being produced by `ProtobufExportOutputSerializer`?
```

Reply:
We have not actually committed to this yet. I do not want to call that out at all.

Reply:
As for ExportOutput holding model_proto... the type's docstring itself says such. The serializer takes the in-memory representation and ... serializes it.

Reply:
> As for ExportOutput holding model_proto... the type's docstring itself says such. The serializer takes the in-memory representation and ... serializes it.

I got a little confused because the suggested serializer implementation is just a "write to disk something already proto-serialized by another component". It doesn't plumb the real serialization through torchscript/onnxscript. In this sense, the serializer is just a "disk-writer" of an already serialized model. That could be seen as an unnecessary layer and my comment was in the sense to clear the air on the what could be seen as an over design thing.

Have we not commited to a generic IR that is further serialized to protobuf (or whatever), we could completely eliminate the `ExportOutputSerializer` concept and do a simple `destination.write(self.model_proto.SerializeToString())` as `ExportOutput.save(self, *, destination: Union[str, io.BufferedIOBase])`. `ExportOutputSerializer` would only be introduced if we ever decide to support anything other than protobuf or decided that `ExportOutput` would not receive a `model_proto` as input.

Makes sense?